include/boost/url/experimental/detail/formatter.hpp

```diff
@@ -0,0 +1,267 @@
+}
+
+template<class CharSet>
+std::size_t
+measure_one(
+    char c,
+    CharSet const& unreserved)
+{
+    // '%' must be reserved
+    BOOST_ASSERT(! unreserved('%'));
+    if(unreserved(c))
+        return 1;
+    return 3;
+}
+
+inline
+char const*
+parse_empty_modifiers(
+    char const* it,
+    char const* end)
+{
+    if (it != end &&
+        *it != '}')
+    {
+        urls::detail::throw_invalid_argument();
+    }
+    return end;
+}
+
+template <>
+struct formatter<char>
+{
+public:
+    char const*
+    parse(
+        char const* it,
+        char const* end) const
+    {
+        return parse_empty_modifiers(it, end);
+    }
+
+    template<class CharSet>
+    std::size_t
+    measure(
+        char c,
+        CharSet const& cs) const {
+        return measure_one(c, cs);
+    }
+
+    template<class CharSet>
+    char*
+    format(char c, char* out, CharSet const& cs) const {
+        encode_one(out, c, cs);
+        return out;
+    }
+};
+
+template <>
+struct formatter<string_view>
+{
+public:
+    char const*
+    parse(
+        char const* it,
+        char const* end) const
+    {
+        return parse_empty_modifiers(it, end);
+    }
+
+    template<class CharSet>
+    std::size_t
+    measure(
+        string_view str,
+        CharSet const& cs) const {
+        return encoded_size(str, cs);
+    }
+
+    template<class CharSet>
+    char*
+    format(string_view str, char* out, CharSet const& cs) const {
+        // unsafe `encode`, assuming `out` has
+        // enough capacity
+        for (char c: str)
+            encode_one(out, c, cs);
+        return out;
+    }
+};
+
+template <>
+struct formatter<const char*>
+    : formatter<string_view>
+{
+};
+
+template <>
+struct formatter<std::string>
+    : formatter<string_view>
+{
+};
+
+template <>
+struct formatter<unsigned long long int>
+{
+public:
+    char const*
+    parse(
+        char const* it,
+        char const* end) const
+    {
+        return parse_empty_modifiers(it, end);
+    }
+
+    template<class CharSet>
+    std::size_t
+    measure(
+        unsigned long long int v,
+        CharSet const& cs) const {
+        std::size_t n = 0;
+        while (v > 0)
+        {
+            int d = v % 10;
+            v /= 10;
+            n += measure_one('0' + d, cs);
+        }
+        return n;
+    }
+
+    template<class CharSet>
+    char*
+    format(unsigned long long int v, char* out, CharSet const& cs) const
+    {
+        std::size_t v0 = v;
+        std::size_t n = 0;
+        std::size_t p = 1;
+        while (v > 0)
+        {
+            if (v >= 10)
+                p *= 10;
+            v /= 10;
+            ++n;
+        }
+        static constexpr auto m =
+            std::numeric_limits<
+                unsigned long long int>::digits10;
+        BOOST_ASSERT(n <= m + 1);
+        ignore_unused(m);
+        v = v0;
+        while (n)
+        {
+            int d = v / p;
+            encode_one(out, '0' + d, cs);
```

Comment:
but we don't need the character set here? for digits?

Reply:
The charset is a parameter because we don't know in what component the string will be formatted. 

Maybe we could consider we don't need to generalize the algorithm that much because digits are accepted unencoded in all components so we could bypass the `cs` here and the same could be said about other combinations of component/type/charset. 

I'm not sure removing the general algorithm is worth it though. The code might become more difficult to follow without a big benefit.